### PR TITLE
feat(twap): pruning keep the newest record under keep threshold

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #2518

## What is the purpose of the change

Modify TWAP pruning to keep the newest record younger than block time - keep period.

Please see #2518 for details.

Additionally, it switched assertions in `api_test.go` from asserting on error strings to actual errors. This required defining several custom errors.

## Brief Changelog

- Modify pruning logic
- Modify existing table-driven tests to account for the change
- Add more tests for getting TWAP from times earlier than block time - keep period
- Modify hook test

## Testing and Verifying

This change added tests mentioned in changelog

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/twap/api_test.go [hunk 1]
```diff
@@ -238,9 +237,156 @@ func (s *TestSuite) TestGetArithmeticTwap() {
 	for name, test := range tests {
 		s.Run(name, func() {
 			s.SetupTest()
-			for _, record := range test.recordsToSet {
-				s.twapkeeper.StoreNewRecord(s.Ctx, record)
```
Comment (p0mvn) [inline x/twap/api_test.go:242]: Note to reviewers: this is the only change in this test fixture (converting loop to using `preSetRecords` helper

File: x/twap/logic_test.go [hunk 2]
```diff
@@ -210,8 +210,9 @@ func TestComputeArithmeticTwap(t *testing.T) {
 	}
 }
 
-// TestPruneRecords tests that all twap records earlier than
-// current block time - RecordHistoryKeepPeriod are pruned from the store.
+// TestPruneRecords tests that twap records earlier than
+// current block time - RecordHistoryKeepPeriod are pruned from the store
+// while keeping the newet record before the above time threshold.
```
Comment (p0mvn) [inline x/twap/logic_test.go:215]:
```suggestion
// while keeping the newest record before the above time threshold.
```

**Review by czarcas7ic — APPROVED** [review]
Well done, all the additional twap logic LGTM so I'm approving! 🚀  

Just have small nits that can be addressed at leisure

File: x/twap/api_test.go [hunk 3]
```diff
@@ -238,9 +237,158 @@ func (s *TestSuite) TestGetArithmeticTwap() {
...
+			var twap sdk.Dec
+			var err error
+
+			twap, err = s.twapkeeper.GetArithmeticTwap(s.Ctx, test.input.poolId,
+				test.input.quoteAssetDenom, test.input.baseAssetDenom,
+				test.input.startTime, test.input.endTime)
+
+			if test.expErrorStr != "" {
+				s.Require().Error(err)
+				s.Require().Contains(err.Error(), test.expErrorStr)
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/twap/api_test.go:252]: Why do we care which error was returned?

File: x/twap/store.go [hunk 4]
```diff
@@ -50,10 +50,27 @@ func (k Keeper) storeHistoricalTWAP(ctx sdk.Context, twap types.TwapRecord) {
...
+// - Suppose pruning param -48 hour
+// - Suppose swaps at -50 hour, -1hour
+// - A prune would leave us with only one record at -1 hour, and we are not able to get twaps from the
+// [-48 hour, -1 hour] time range.
+func (k Keeper) pruneRecordsBeforeTimeButNewest(ctx sdk.Context, lastTime time.Time) error {
 	store := ctx.KVStore(k.storeKey)
-	iter := store.Iterator([]byte(types.HistoricalTWAPTimeIndexPrefix), types.FormatHistoricalTimeIndexTWAPKey(lastTime, 0, "", ""))
+
+	// Reverse iterator guarantees that we get the newest records first.
+	iter := store.ReverseIterator([]byte(types.HistoricalTWAPTimeIndexPrefix), types.FormatHistoricalTimeIndexTWAPKey(lastTime, 0, "", ""))
```
Comment (ValarDragon) [inline x/twap/store.go:66]: I don't think this is right, to be right we'd have to have skipping logic per pool. This only maintains skipping logic for pool 0

File: x/twap/api_test.go [hunk 5]
```diff
@@ -238,9 +237,158 @@ func (s *TestSuite) TestGetArithmeticTwap() {
...
+		"(1 record younger than keep threshold); to now; ctxTime = start time = end time = after keep threshold": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      oneHourAfterKeepThreshold,
+			input:        makeSimpleTwapInput(oneHourAfterKeepThreshold, oneHourAfterKeepThreshold, quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record younger than keep threshold); with end time; ctxTime = start time = end time = after keep threshold - 1ns": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      oneHourAfterKeepThreshold,
+			input:        makeSimpleTwapInput(oneHourAfterKeepThreshold.Add(-time.Millisecond), oneHourAfterKeepThreshold.Add(-time.Millisecond), quoteAssetA),
```
Comment (ValarDragon) [inline x/twap/api_test.go:320]: I feel like we could probably add some of these table cases more clearly, via a for loop. Don't need to change if you don't want to, but I think could make adding more cases easier

**Review by ValarDragon — CHANGES_REQUESTED** [review]
I don't think the pruning logic works correctly, for keeping the latest record per pool

File: x/twap/api_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/twap/api_test.go:252]: @ValarDragon here is a recent example https://github.com/osmosis-labs/osmosis/pull/2536#discussion_r958758612

File: x/twap/store.go (hunk 4 above)
Comment (p0mvn) [reply x/twap/store.go:66]: Yeah that's a huge oversight on my part, thanks for catching that. Fixed now and refactored tests to be more thorough and capture coverage across pools

File: x/twap/store_test.go [hunk 6]
```diff
@@ -245,11 +246,18 @@ func (s *TestSuite) TestPruneRecordsBeforeTime() {
...
 		},
-		"base time minus 1, 2 records before (deleted), 1 records at (deleted), 1 records after (not deleted)": {
+		"base time minus 1, 1 record before (not deleted, newest kept), 1 records at (not deleted), 2 records after (not deleted)": {
 			recordsToPreSet: allTestRecords,
 
 			beforeTime: tMin1Record.Time,
 
+			expectedKeptRecords: []types.TwapRecord{tMin2Record, tMin1Record, baseRecord, tPlus1Record},
+		},
+		"base time, 2 records before (one deleted, newest kept), 1 records at (not deleted), 1 records after (not deleted)": {
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/twap/store_test.go]: Tests were changed so this is now obsolete

File: x/twap/store_test.go [hunk 7]
```diff
@@ -245,11 +246,18 @@ func (s *TestSuite) TestPruneRecordsBeforeTime() {
 
 			expectedKeptRecords: []types.TwapRecord{tPlus1Record},
 		},
-		"base time minus 1, 2 records before (deleted), 1 records at (deleted), 1 records after (not deleted)": {
+		"base time minus 1, 1 record before (not deleted, newest kept), 1 records at (not deleted), 2 records after (not deleted)": {
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/twap/store_test.go]: Changed tests

File: x/twap/api_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/twap/api_test.go]: Fixed all errors I observed in the file. Generally, let's create separate issues for such things. I think it's best to keep the scope focused on the actual change instead of fixing drive-by problems

File: x/twap/api_test.go (hunk 5 above)
Comment (p0mvn) [reply x/twap/api_test.go:318]: I explored this idea [here](https://github.com/osmosis-labs/osmosis/pull/2543/commits/e8e5084271c680ae2310f830ce4b8df1fc565968)

I don't think this is easily doable because if I generalize test cases to always subtract from end time to trigger the "with end time" path, the expected value of `expTwap` changes.

I cannot generalize by instead modifying `ctxTime` to trigger the "with end time" path because test cases' logic sometimes depends on it.

Please let me know if you see something I'm not, and I can change this

**Comment by p0mvn** [comment]
All comments are addressed, please take a look

File: x/twap/api_test.go [hunk 8]
```diff
@@ -213,34 +213,181 @@ func (s *TestSuite) TestGetArithmeticTwap() {
...
+		"(1 record at keep threshold); with end time; ctxTime = base keep threshold, start time = base time - 1ms (source of error); end time = base keep threshold - ms; error": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTimePlusKeepPeriod,
+			input:        makeSimpleTwapInput(baseTime.Add(-time.Millisecond), baseTimePlusKeepPeriod.Add(-time.Millisecond), quoteAssetA),
+			expectError:  twap.TimeTooOldError{Time: baseTime.Add(-time.Millisecond)},
+		},
+		"(2 records); to now; with one directly at threshold, interpolated": {
+			recordsToSet: []types.TwapRecord{baseRecord, recordBeforeKeepThreshold},
+			ctxTime:      baseTimePlusKeepPeriod,
+			input:        makeSimpleTwapInput(baseTime, baseTimePlusKeepPeriod, quoteAssetA),
```
Comment (mattverse) [inline x/twap/api_test.go:360]: Seeing all these test cases, Im thinking if there would be a way to declare all these test cases as variables(including the expected TWAP) in the future

File: x/twap/listeners_test.go [hunk 9]
```diff
@@ -201,17 +201,31 @@ func (s *TestSuite) TestEndBlock() {
...
+
+	// Assume some time has passed and new record created.
+	s.Ctx = s.Ctx.WithBlockTime(tPlus10sp5Record.Time)
+	newestRecord := tPlus10sp5Record
+
+	s.twapkeeper.StoreNewRecord(s.Ctx, newestRecord)
+
+	twapsBeforeEpoch, err := s.twapkeeper.GetAllHistoricalTimeIndexedTWAPs(s.Ctx)
 	s.Require().NoError(err)
+
```
Comment (mattverse) [inline x/twap/listeners_test.go:221]:
```suggestion
```

File: x/twap/store_test.go [hunk 10]
```diff
@@ -201,70 +201,177 @@ func (s *TestSuite) TestGetRecordAtOrBeforeTime() {
...
+func (s *TestSuite) TestPruneRecordsBeforeTimeButNewest() {
+	// These are manually created to be able to refer to them by name
+	// for convenience.
+	pool1Min2SBaseMs, pool2Min1SBaseMs, pool3BaseSecBaseMs, pool4Plus1SBaseMs := s.createTestRecordsFromTime(baseTime)
+	pool1Min2SMin1Ms, pool2Min1SMin1Ms, pool3BaseSecMin1Ms, pool4Plus1SMin1Ms := s.createTestRecordsFromTime(baseTime.Add(-time.Millisecond))
+	pool1Min2SMin2Ms, pool2Min1SMin2Ms, pool3BaseSecMin2Ms, pool4Plus1SMin2Ms := s.createTestRecordsFromTime(baseTime.Add(2 * -time.Millisecond))
+	pool1Min2SMin3Ms, pool2Min1SMin3Ms, pool3BaseSecMin3Ms, pool4Plus1SMin3Ms := s.createTestRecordsFromTime(baseTime.Add(3 * -time.Millisecond))
+
+	// Create records that match times 1:1 with other pools but have a different pool ID.
+	pool5Min2SBaseMs, pool5Min1SBaseMs, pool5BaseSecBaseMs, pool5Plus1SBaseMs := s.createTestRecordsFromTime(baseTime)
```
Comment (mattverse) [inline x/twap/store_test.go:215]: im wondering if there's reason we need to test with so many pools, since I think max 3 pools would suffice testing any case possible with multiple pools

**Review by mattverse — COMMENTED** [review]
It might just be a personal opinion, but tests in `store_test.go` seems a little bit hard to follow,(test case names are not quite intuitive on what exactly we want to test in that specific case). Other than that main logic lgtm!

File: x/twap/store.go [hunk 11]
```diff
@@ -12,6 +12,25 @@ import (
...
+}
+
+type twapNotFoundError struct {
+	Asset0Denom string
+	Asset1Denom string
+}
+
+func (e twapNotFoundError) Error() string {
+	return fmt.Sprintf("TWAP not found, but there are other twaps available for this time."+
+		" Please make sure tha asset0denom and asset1denom (%s, %s) are correct, and in order (asset0 > asset1)?", e.Asset0Denom, e.Asset1Denom)
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/twap/store.go:31]: We can address this separately if desired. This is simply extracting errors to be reused in assertions in tests

File: x/twap/store.go [hunk 12]
```diff
@@ -50,15 +69,36 @@ func (k Keeper) storeHistoricalTWAP(ctx sdk.Context, twap types.TwapRecord) {
...
+// [-48 hour, -1 hour] time range.
+func (k Keeper) pruneRecordsBeforeTimeButNewest(ctx sdk.Context, lastTime time.Time) error {
 	store := ctx.KVStore(k.storeKey)
-	iter := store.Iterator([]byte(types.HistoricalTWAPTimeIndexPrefix), types.FormatHistoricalTimeIndexTWAPKey(lastTime, 0, "", ""))
+
+	// Reverse iterator guarantees that we get the newest records first.
+	iter := store.ReverseIterator([]byte(types.HistoricalTWAPTimeIndexPrefix), types.FormatHistoricalTimeIndexTWAPKey(lastTime, 0, "", ""))
 	defer iter.Close()
+
+	seenPools := map[uint64]struct{}{}
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/twap/store.go:88]: The goal is to have a set. If a boolean was here, it would be wasting space. `struct{}{}` allocates 0 bytes for each key and `map[uint64]struct{}{}` is idiomatic with a set in Go

File: x/twap/api_test.go (hunk 8 above)
Comment (p0mvn) [reply x/twap/api_test.go:360]: These are inputs though and they are constructed based on the logic / setup of the test case. I don't think we should be extracting these since the setup logic varies form test case to test case

File: x/twap/store_test.go (hunk 10 above)
Comment (p0mvn) [reply x/twap/store_test.go:215]: They all start at the different base times:
https://github.com/osmosis-labs/osmosis/blob/2c7649f4ed193e05ae921c65cb2a037d9c3412d3/x/twap/store_test.go#L209-L212


The goal is not only to test the inter-pool correctness but also how pruning works with respect to different times

File: x/twap/api.go [hunk 13]
```diff
@@ -9,6 +9,26 @@ import (
...
+
+type startTimeAfterEndTimeError struct {
+	StartTime time.Time
+	EndTime   time.Time
+}
+
+func (e startTimeAfterEndTimeError) Error() string {
+	return fmt.Sprintf("called GetArithmeticTwap with a start time that is after the end time."+
+		" (start time %s, end time %s)", e.StartTime, e.EndTime)
+}
```
Comment (ValarDragon) [inline x/twap/api.go:30]: Lets move these to bottom of file?

File: x/twap/store.go [hunk 14]
```diff
@@ -50,15 +69,36 @@ func (k Keeper) storeHistoricalTWAP(ctx sdk.Context, twap types.TwapRecord) {
...
+// For example:
+// - Suppose pruning param -48 hour
+// - Suppose swaps at -50 hour, -1hour
+// - A prune would leave us with only one record at -1 hour, and we are not able to get twaps from the
+// [-48 hour, -1 hour] time range.
+func (k Keeper) pruneRecordsBeforeTimeButNewest(ctx sdk.Context, lastKeptTime time.Time) error {
 	store := ctx.KVStore(k.storeKey)
-	iter := store.Iterator([]byte(types.HistoricalTWAPTimeIndexPrefix), types.FormatHistoricalTimeIndexTWAPKey(lastTime, 0, "", ""))
+
+	// Reverse iterator guarantees that we get the newest records first.
```
Comment (ValarDragon) [inline x/twap/store.go:84]:
```suggestion
	// Reverse iterator guarantees that we iterate through the newest per pool first.
	// Due to how its indexed, we will only iterate times starting from lastKeptTime
```

File: x/twap/store.go [hunk 15]
```diff
@@ -50,15 +69,36 @@ func (k Keeper) storeHistoricalTWAP(ctx sdk.Context, twap types.TwapRecord) {
...
+// pruneRecordsBeforeTimeButNewest prunes all records for each pool before the given time but the newest
+// record. The reason for preserving at least one record earlier than the keep period is
+// to ensure that we have a record to interpolate from in case there is only one or no records
+// within the keep period.
+// For example:
+// - Suppose pruning param -48 hour
+// - Suppose swaps at -50 hour, -1hour
+// - A prune would leave us with only one record at -1 hour, and we are not able to get twaps from the
+// [-48 hour, -1 hour] time range.
+func (k Keeper) pruneRecordsBeforeTimeButNewest(ctx sdk.Context, lastKeptTime time.Time) error {
```
Comment (ValarDragon) [inline x/twap/store.go:81]: Should we rename lastKeptTime to firstRemoveTime due to how inclusivity works in the iterator?

**Comment by ValarDragon** [comment]
Logic LGTM, have not yet reviewed tests!

File: x/twap/api.go (hunk 13 above)
Comment (p0mvn) [reply x/twap/api.go]: Moved to `types/errros,go`

File: x/twap/store.go (hunk 15 above)
Comment (ValarDragon) [reply x/twap/store.go:81]: ah that makes sense

**Review by ValarDragon — APPROVED** [review]
Nice job on the feature + thorough tests!

I somewhat feel like we went overboard with long hand written test cases here though, versus having a lot of these vectors made via loops, randomization, or more segmented.
